File: leetcode/js/sqrtx.js
/**
 * @param {number} x
 * @return {number}
 */
var mySqrt = function (x) {
  if (x <= 1) {
    return x;
  }

  let left = 0;
  let right = Math.floor(x / 2);

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    const pow = mid * mid;

    if (pow === x) {
      right = mid;
      break;
    } else if (pow < x) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return right;
};

File: leetcode/rust/sqrtx.rs
impl Solution {
    pub fn my_sqrt(x: i32) -> i32 {
        use std::cmp::Ordering;

        if x <= 1 {
            return x;
        }

        let x = x as u64;
        let mut left = 0;
        let mut right = x / 2;

        while left <= right {
            let mid = (left + right) / 2;
            let pow = mid * mid;

            match pow.cmp(&x) {
                Ordering::Less => {
                    left = mid + 1;
                }
                Ordering::Greater => {
                    right = mid - 1;
                }
                Ordering::Equal => {
                    right = mid;
                    break;
                }
            }
        }

        right as i32
    }
}
